src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -274,58 +276,88 @@ bool AbstractLQPNode::has_matching_unique_constraint(const ExpressionUnorderedSe
+             "Did not expect to generate non-trivial FD from unique constraints.");
     }
   }
+  if (generated_fds.empty()) return non_trivial_fds;
 
-  for (const auto& fd : fds_in) {
-    // Check whether the determinants are a subset of the current node's non-nullable output expressions
-    if (std::any_of(fd.determinants.cbegin(), fd.determinants.cend(),
-                    [&output_expressions_non_nullable](const auto& expression) {
-                      return !output_expressions_non_nullable.contains(expression);
-                    })) {
-      continue;
-    }
-
-    // We do not check the columns of the FD's dependents set since they are allowed to be nullable.
+  // (3) Create single output vector
+  auto output_fds = non_trivial_fds;
+  output_fds.reserve(non_trivial_fds.size() + generated_fds.size());
+  std::move(generated_fds.begin(), generated_fds.end(), std::back_inserter(output_fds));
+  return output_fds;
+}
 
-    // FD remains valid, so add it to the output vector
-    fds_out.push_back(fd);
+void AbstractLQPNode::_remove_invalid_fds(std::vector<FunctionalDependency>& fds) const {
+  if (fds.empty()) return;
+  const auto& output_expressions = this->output_expressions();
+  const auto& output_expressions_set = ExpressionUnorderedSet{output_expressions.cbegin(), output_expressions.cend()};
+  fds.erase(std::remove_if(fds.begin(), fds.end(),
+                           [this, &output_expressions_set](const auto& fd) {
+                             /**
+                              * Remove FDs with determinant expressions that are
+                              *  a) not part of the node's output expressions
+                              *  b) are nullable
+                              */
+                             for (const auto& fd_determinant_expression : fd.determinants) {
+                               if (!output_expressions_set.contains(fd_determinant_expression) ||
+                                   is_column_nullable(get_column_id(*fd_determinant_expression))) {
+                                 return true;
+                               }
+                             }
+
+                             // Remove dependents that are not part of the node's output expressions
+                             auto part_of_output_expressions =
+                                 [&output_expressions_set](const auto& fd_dependent_expression) {
+                                   return !output_expressions_set.contains(fd_dependent_expression);
+                             };
+                             std::erase_if(fd.dependents, part_of_output_expressions);
+
+                             // If there are no dependents left, we can discard the FD altogether
+                             return fd.dependents.size() == 0;
+                           }),
+            fds.end());
+}
+
+std::vector<FunctionalDependency> AbstractLQPNode::non_trivial_functional_dependencies() const {
```

Comment:
I can replace this with `return {};` without any tests failing. :/

We should have a handful of more complex LQPs in the tests.

Reply:
I have added several tests for
* AggregateNode
* ProjectionNode
* JoinNode

Not sure though, if we need another one.